### PR TITLE
Implement object checkout in PrepareCheckout

### PR DESCRIPTION
**Pull request by Fernthedev**
In continuation of #1403, I implemented `with_object_id` to allow checking out specific commits such as is the case in submodules. 

This currently does not include any test cases as of writing this. [However, I did make sure this worked from a high level in my project](https://github.com/QuestPackageManager/QPM.CLI/blob/acf6561f16fd700d012295e2bdf735221dfc8399/src/utils/git.rs#L258-L260)

Though this does not properly update the `HEAD` and may not clone properly. Hoping I can fix it by using `util::update_head` and refactoring it to use `oid`, though I leave that decision up to you if you have a better idea.

Apologies for the lack of tests as well.

Tasks:
- [x]  Checkout tree associated with the `oid` of the revision used.
- [ ] Update the `HEAD` with `util::update_head` to use the correct detached state.
- [ ] Add tests ensuring this works as intended
-   [ ] Shallow clone with specific revision, then checkout.
-   [ ] Standard clone while checking out specific revision

### CODE REVIEW

File: gix/src/clone/checkout.rs [hunk 1]
```diff
@@ -60,6 +61,54 @@ pub mod main_worktree {
 
     /// Modification
     impl PrepareCheckout {
+        /// Set the `name` of the reference to check out, instead of the remote `HEAD`.
+        /// If `None`, the `HEAD` will be used, which is the default.
+        ///
+        /// Note that `name` should be a partial name like `main` or `feat/one`, but can be a full ref name.
+        /// If a branch on the remote matches, it will automatically be retrieved even without a refspec.
+        pub fn with_ref_name<'a, Name, E>(self, ref_name: Option<Name>) -> Result<PrepareCheckout, Error>
```
Comment (Byron) [inline gix/src/clone/checkout.rs:69]: These three methods can be a single one if a `revspec` is used, see [rev_parse()](https://docs.rs/gix/latest/gix/struct.Repository.html#method.rev_parse) for details.

**Review by Byron — COMMENTED** [review]
Thanks for contributing!

Maybe you could add a task list to communicate what you think is missing to make this PR reviewable.

**Comment by Fernthedev** [comment]
Alright, I added some tasks as requested though the extent of my knowledge with this process is limited so there might be more things I'm not aware of.

**Comment by Byron** [comment]
Thanks for giving this a shot! As the new year is looming, I am closing PRs that can't move and that I can't take over. Thus, sadly, this one is among them but at the same time I encourage to reopen it feels right.
Thanks for your understanding.

**Pull request closed without merging**